### ISSUE TITLE
index_select performance

### ISSUE THREAD

**Issue by ngimel** [pytorch/pytorch#151612]
According to my cursory benchmarking, `gather` is always better than `index_select` and it is strictly more expressive
```
def comp_index_select(M,N,K, dim):
    def gather(x, dim, ind):
        shape_ind = [1] * x.ndim
        shape_ind[dim] = ind.shape[0]
        shape_out = list(x.shape)
        shape_out[dim] = ind.shape[0]
        ind = ind.view(shape_ind).expand(shape_out)
        return torch.gather(x, dim=dim, index=ind)

    src = torch.randn(M,K, device="cuda", dtype=torch.bfloat16)
    max_ind = src.shape[dim]
    num_ind = N
    indices = torch.randint(max_ind, (num_ind,), device="cuda")

    out1 = torch.index_select(src, dim=dim, index=indices)
    out2 = gather(src, dim, indices)
    torch.testing.assert_close(out1, out2)
    from torch.utils.benchmark import Timer
    t = Timer(stmt="torch.index_select(src, dim=dim, index=indices)", globals={"src": src, "dim": dim, "indices": indices})
    print("Torch time", t.blocked_autorange().mean)
    t = Timer(stmt="gather(src, dim, indices)", globals={"src": src, "dim": dim, "indices": indices, "gather": gather})
    print("Torch time", t.blocked_autorange().mean)

comp_index_select(4096, 4096, 5120, 0)
comp_index_select(4096, 4096, 5124, 0)
comp_index_select(4096, 4096, 5120, 1)
```
```
index_select time 0.00013564855395816267
gather time 3.6312990123406055e-05
index_select time 0.00013131444132886828
gather time 6.217161598615348e-05
index_select time 0.00012328573293052616
gather time 6.178460847586394e-05```
```
If more comprehensive benchmarking confirms that to be true, we should expand `gather` coverage to have all the same dtypes as `index_select`, call into `gather` from `index_select` and remove `index_select` kernels. 

cc @msaroufim @jerryzh168

**Comment by isuruf** [comment]
There is a faster code path in gather when it's the first index and the source is contiguous and all the dims are divisible by 16. This explains the almost 4 times speedup in `comp_index_select(4096, 4096, 5120, 0)`. Looking at why the others have a 2 times speedup.

**Comment by ngimel** [comment]
Faster code path was merged yesterday and doesn't exist in nightlies yet, but even without faster codepath it seems like gather is faster than index_select.

**Comment by ngimel** [comment]
@vince62s I was surprised by the difference between `custom indexing cuda` and `pytorch cuda` because they literally are running the same kernel, turns out in the first benchmark you are paying for loading the indexing kernel to the device (it's lazy loaded the first time it's used), if you do a warmup run then those times become pretty close (you start measuring some cpu overhead at this point because your sizes are so small, so custom_indexint_cuda is still slightly slower because it is doing more python overhead)

**Comment by isuruf** [comment]
>  In any case, we shouldn't have different kernels for ops that do literally the same.

True. We can have `index_select` call into `gather`. For small indices for example `comp_index_select(4096, 8, 5120, 0)` the `index_select` kernel seems to be faster.

**Comment by vince62s** [comment]
Just out of curiosity, does `gather(src, 0, indices)` call the same kernel as `src[indices]` ?

actually a syntax like src[:, indices] when dim has to be second dim, would be nice to have.

**Comment by isuruf** [comment]
> Just out of curiosity, does gather(src, 0, indices) call the same kernel as src[indices] ?

No, those are two different kernels.

**Comment by vince62s** [comment]
> `src[:, indices] ` is supported, is there something else you had in mind?

hmm has this always been the case? because if so: 1) I don't see the need for index_select() and 2) I don't recall when I did the snippet above (several months ago) why I had to do the custom_index_select()

But I might be wrong.

**Comment by isuruf** [comment]
> hmm has this always been the case? because if so: 1) I don't see the need for index_select() and 2) I don't recall when I did the snippet above (several months ago) why I had to do the custom_index_select()

Yes, it has been supported for a while. `index_select` is a convenience to write `src[:, indices]` for arbitrary dimensions.

**Comment by isuruf** [comment]
See https://github.com/pytorch/pytorch/pull/151715

**Comment by ngimel** [comment]
`index_select` is a legacy function that predates advanced indexing support. `src[:, indices]` has been supported for several years.

**Comment by ngimel** [comment]
For completeness, now `gather(src, 0, indices) call the same kernel as src[indices] ` for contiguous aligned slices are routed to the same fast kernel.
